lib/azure/armrest/model/hash_model.rb

```diff
@@ -0,0 +1,140 @@
+require 'active_support/core_ext/string/inflections'
+
+module Azure
+  module Armrest
+    class HashModel
+      class << self
+        private
+
+        # Convert a key to a model class attribute which must be a qualified ruby method name
+        # key is from the original JSON or Hash object
+        #
+        # This is the default implementation, to be overridden by subclasses.
+        #
+        def key_to_attribute(key)
+          key.to_s.underscore
+        end
+
+        # Should setter methods be automatically created?
+        #
+        # Default to true, to be overridden by subclasses.
+        #
+        def generate_writer?
+          true
+        end
+
+        # Should values of this attribute be excluded from converting to model objects
+        #
+        # Default to false, to be overridden by subclasses.
+        #
+        def attr_excluded?(_attr_name)
+          false
+        end
+
+        # Declare attributes that need to be converted to Model
+        def attr_model(*attrs)
+          # Merge the declared attributes to the modelized list
+          @modelized_list = modelized_list | Set.new(attrs.map(&:to_s))
+        end
+
+        # Return the list of modelized attributes
+        def modelized_list
+          @modelized_list ||= Set.new
+        end
+
+        def convert_value(key, value, parent)
+          method = key_to_attribute(key)
+          new_val =
+            if attr_excluded?(method)
+              value
+            else
+              # Must deal with nested models
+              case value
+              when Array
+                value.collect do |elem|
+                  elem.kind_of?(Hash) ? parent.send(:hash_to_model, method.camelize.singularize, elem) : elem
```

Comment:
Would possibly a cached Hash for `key_to_attribute` and `method.camelize_sigularize` be helpful here?  I realize there would be thread safety issues with that as well that would need to be addressed, but it would avoid the costly (and repetitive) operations of doing these methods multiple times for every object instantiated.